.grit/patterns/hathora_to_new_ts_sdk.md

```diff
@@ -0,0 +1,226 @@
+    // TODO how do we match and use rest-of-args without so much copypasta?
+    if ($args <: [$app_id, $room_id, $req, $opts]) {
+      $use_args = join(list=[$req, $app_id, $room_id, $opts], separator=", ")
+    } else if ($args <: [$app_id, $room_id, $req]) {
+      $use_args = join(list=[$req, $room_id, $app_id], separator=", ")
+    } else {
+      $use_args = $args
+    }
+  } else if ($method <: `loginGoogle`) {
+    if ($args <: [$app_id, $req, $opts]) {
+      $use_args = join(list=[$req, $app_id, $opts], separator=", ")
+    } else if ($args <: [$app_id, $req]) {
+      $use_args = join(list=[$req, $app_id], separator=", ")
+    } else {
+      $use_args = $args
+    }
+  } else {
+    $use_args = join(list=$args, separator=", ")
+  },
+  return $use_args
+}
+
+pattern rewrite_method_calls($resource, $callee) {
+  maybe any {
+    // TODO more elegant handling of await-or-not, maybe via regex? better idea?
+    contains bubble($resource, $callee) `await $callee.$method($args)` as $body where {
+      $suffix = deprecated_suffix(resource=$resource, method=$method),
+      // TODO finish rearranging methods
+      // TODO generalize arg-rearrangement for the non-await case, since they're the same.
+      $use_args = reorder_args($method, $args),
+      $unwrap_at = wrapper_key(method=$method),
+      $body => `(await $callee.$[method]$[suffix]($use_args)).$unwrap_at`
+    },
+    contains bubble($resource, $callee) `$callee.$method($args)` as $body where {
+      $suffix = deprecated_suffix(resource=$resource, method=$method),
+      $use_args = reorder_args($method, $args),
+      $body => `$callee.$[method]$[suffix]($use_args)`
+    }
+  }
+}
+
+pattern generic_replace_resource($resource) {
+  and {
+      replace_resource_import($resource),
+    // TODO what do we do about hygiene when we import names like `sdk`?
+    // do we have to worry about shadowing?
+      maybe any {
+        $member_name = sdk_member_name($resource),      
+        and {
+          contains `const $binding = new $resource()` => `const $binding = sdk.$member_name;`, 
+          rewrite_method_calls(resource=$resource, callee=$binding),
+        },
+        contains type_annotation(type=$resource) where {
+          $source = resource_type_import_source($resource),
+          $substitute = resource_type_substitute($resource),
+          $substitute <: ensure_import_from(source=$source)
+      } => `: $substitute`
+    }
+  }
+}
+
+// v1 functions return an extra layer of indirection now, so eg a login response is at `.loginResponse`
+pattern replace_inside_functions($client_type) {
+  function_declaration(parameters=$params, body=$body) where {
+    // TODO critically, this depends on types. if the target project is pure js, we're SoL.
+    // One option is to just assume any name calling a given method like `getConnectionInfo` is using Hathora.
+    // Not every method name may be that distinctive, and in any real codebase it could be insanely noisy.
```

Comment:
Just follow this method. Our experience has been that false positives are a lot less common than you might expect and we can easily add some side conditions (like checking the file actually imports the SDK) to get around them.